### ISSUE TITLE
Editable Table Example

### ISSUE THREAD

**Issue by AndyObtiva** [kojix2/LibUI#25]
Hi,

Thank you for providing such a useful Ruby gem and for adding [Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui) to the list of supported LibUI DSLs.

I wanted to note that I just released initial support for the `table` control, supporting only `text_column`s for the time being.

Your implementation of [examples/basic_table.rb](https://github.com/kojix2/LibUI/blob/main/examples/basic_table.rb) was a very helpful reference for my DSL reimplementation.

[Glimmer DSL for LibUI](https://github.com/AndyObtiva/glimmer-dsl-libui) [examples/basic_table.rb](https://github.com/AndyObtiva/glimmer-dsl-libui/tree/master#basic-table)
```ruby
require 'glimmer-dsl-libui'

include Glimmer

data = [
  %w[cat meow],
  %w[dog woof],
  %w[chicken cock-a-doodle-doo],
  %w[hourse neigh],
  %w[cow moo]
]

window('Animal sounds', 300, 200) {
  horizontal_box {
    table {
      text_column('Animal')
      text_column('Description')

      cell_rows data
    }
  }
  
  on_closing do
    puts 'Bye Bye'
  end
}.show
```

I aim to support different types of table columns eventually like `image_column`, and I am able to pursue this effort on my own.

That said, I also plan to support editable `table`s, and this is where I need help. I was wondering if you could provide `examples/editable_table.rb` example similar to `examples/basic_table.rb` that simply enables table editing.

What I understand so far on my own is that these two lines of code need to have their last argument changed from `-1` to `-2` to enable editing (I tried it successfully):
```ruby
UI.table_append_text_column(table, 'Animal', 0, -1)
UI.table_append_text_column(table, 'Description', 1, -1)
```
becomes
```ruby
UI.table_append_text_column(table, 'Animal', 0, -2)
UI.table_append_text_column(table, 'Description', 1, -2)
```

Next, to persist changes of edited values, this block of code needs to do extra work:
```ruby
model_handler.SetCellValue = rbcallback(0, [0]) {}
```

That is exactly where I need your insight. I do not know Fiddle at all (a big reason why I highly appreciate your Ruby gem) so I do not understand how to hook into [libui](https://github.com/andlabs/libui) correctly to receive updated cell values and save them into the `data` array (e.g. via some code like `data[row][column] = value`)

If you could help me by producing the suggested `examples/editable_table.rb` example, I would be very grateful!

Best regards,

Andy

**Comment by kojix2** [comment]
Hi @AndyObtiva 

As you said, it looks like you need to write `SetCellValue`.
I think the following example will work.

```ruby
model_handler.SetCellValue = rbcallback(0, [1, 1, 4, 4, 1]) do |_, _, row, column, val|
  data[row][column] = UI.table_value_string(val).to_s
end
```

Thank you for creating glimmer-dsl-libui!

---

`rbcallback(0, [1, 1, 4, 4, 1])`

```ruby
require 'fiddle'
Fiddle.constants.each do |c|
  if c.to_s.start_with? "TYPE_"
    puts "#{c}\t#{Fiddle.const_get(c)}"
  end
end
```

```
TYPE_VOID       0
TYPE_CHAR       2
TYPE_SHORT      3
TYPE_LONG_LONG  6
TYPE_INT8_T     2
TYPE_INT16_T    3
TYPE_INT32_T    4
TYPE_INT64_T    6
TYPE_FLOAT      7
TYPE_DOUBLE     8
TYPE_VARIADIC   9
TYPE_CONST_STRING       10
TYPE_SIZE_T     -6
TYPE_SSIZE_T    6
TYPE_PTRDIFF_T  6
TYPE_INTPTR_T   6
TYPE_UINTPTR_T  -6
TYPE_LONG       5
TYPE_INT        4
TYPE_VOIDP      1
```

**Comment by AndyObtiva** [comment]
Thank you. The code worked. I'll look into amending it for the different column types in the future.

I just released [Glimmer DSL for LibUI 0.0.19](https://github.com/AndyObtiva/glimmer-dsl-libui) with [examples/editable_table.rb](https://github.com/AndyObtiva/glimmer-dsl-libui#editable-table):

```ruby
require 'glimmer-dsl-libui'

include Glimmer

data = [
  %w[cat meow],
  %w[dog woof],
  %w[chicken cock-a-doodle-doo],
  %w[hourse neigh],
  %w[cow moo]
]

window('Editable animal sounds', 300, 200) {
  horizontal_box {
    table {
      text_column('Animal')
      text_column('Description')

      cell_rows data
      editable true
    }
  }
  
  on_closing do
    puts 'Bye Bye'
  end
}.show
```

I will be providing another version that enables setting the `editable` property on specific column(s) only instead of the entire `table`

Cheers.
